// node-hill-s/dist/src/class/Sanction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sanction = void 0;
const fs_1 = __importDefault(require("fs"));
class Sanction {
    constructor() {
        this.allowedIPs = new Set(["127.0.0.1"]);
        this.bannedIPs = new Set();
        this.disabled = false;
        this.debugLogging = false;
    }
    debugLog(data) {
        if (!this.debugLogging)
            return;
        fs_1.default.appendFile("./sanction.txt", JSON.stringify(data) + "\n", () => { });
    }
    banPlayer(player) {
        if (this.bannedIPs.has(player.socket.IPV4))
            throw "Player is already banned!";
        this.bannedIPs.add(player.socket.IPV4);
        player.kick("You have been banned from the server.");
    }
    banSocket(socket, expirationTime = 1000 * 60 * 60) {
        if (this.disabled)
            return;
        // Player is already sanctioned.
        if (this.bannedIPs.has(socket.IPV4) || this.allowedIPs.has(socket.IPV4))
            return false;
        this.bannedIPs.add(socket.IPV4);
        socket.destroy();
        setTimeout(() => {
            this.bannedIPs.delete(socket.IPV4);
        }, expirationTime);
        console.warn(`[SANCTION] ${socket.IPV4} was automatically banned for suspicious traffic.`);
        return true;
    }
}
exports.Sanction = Sanction;
const sanction = new Sanction();
exports.default = sanction;
