// node-hill-s/dist/src/api/postServer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Game_1 = __importDefault(require("../class/Game"));
let postedOnce = false;
const phin = require("phin")
    .defaults({
    url: "https://sandpile.xyz/api/postServer",
    method: "POST",
    timeout: 12000
});
function checkForBannedUsers(bannedUsers) {
    for (const player of Game_1.default.players) {
        if (bannedUsers.includes(player.userId))
            player.kick("Banned from SandPile.");
    }
}
function initializeSetData(setResponse) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const setData = yield Game_1.default.getSetData(setResponse.set_id);
            Game_1.default.setData = setData;
            console.log(`Posted to: https://www.sandpile.xyz/set/${setResponse.set_id} successfully.`);
            postedOnce = true;
            Game_1.default.emit("setDataLoaded");
        }
        catch (err) {
            Game_1.default.emit("setDataLoaded", err);
        }
    });
}
function postServer() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const postData = {
                "host_key": Game_1.default.serverSettings.hostKey,
                "port": Game_1.default.serverSettings.port,
                "players": Game_1.default.players.map(player => player.validationToken)
            };
            const response = yield phin({ data: postData });
            try {
                const data = JSON.parse(response.body);
                if (data.error) {
                    console.warn("Failure while posting to games page:", JSON.stringify(data.error.message || data));
                    if (data.error.message === "You can only postServer once every minute")
                        return;
                    return process.exit(0);
                }
                else {
                    if (!postedOnce)
                        initializeSetData(data);
                    checkForBannedUsers(data.banned_players);
                    return data;
                }
            }
            catch (err) { } // It was successful (?)
        }
        catch (err) {
            console.warn("Error while posting to games page.");
            console.error(err.stack);
        }
    });
}
exports.default = postServer;
